scalyr_agent/sharded_copying_manager/copying_manager.py

```diff
@@ -837,37 +865,45 @@ def generate_status(self, warn_on_rate_limit=False):
             all_health_check_error_messages = list()
 
-            for api_key_id in sorted(self.__api_keys_worker_pools):
-                worker_pool = self.__api_keys_worker_pools[api_key_id]
-                worker_pool_status = worker_pool.generate_status(
+            api_key_statuses = [
+                self.__api_keys_worker_pools[api_key_id].generate_status(
                     warn_on_rate_limit=warn_on_rate_limit
                 )
-                result.api_key_worker_pools.append(worker_pool_status)
-
-                result.total_errors += worker_pool_status.total_errors
-                result.total_bytes_uploaded += worker_pool_status.total_bytes_uploaded
-
-                # if at least one if failed, the result is failed too.
-                if worker_pool_status.all_responses_successful is not None:
-                    # do not overwrite the result if we already have a bad response.
-                    if all_responses_good is not False:
-                        all_responses_good = worker_pool_status.all_responses_successful
-
-                # if at least one if failed, the result is failed too.
-                if worker_pool_status.all_health_checks_good is not None:
-                    # do not overwrite the result if we already have a bad response.
-                    if all_health_checks_good is not False:
-                        all_health_checks_good = (
-                            worker_pool_status.all_health_checks_good
-                        )
-                        if not worker_pool_status.all_health_checks_good:
-                            all_health_checks_good = False
+                for api_key_id in sorted(self.__api_keys_worker_pools)
+            ]
+
+            result.api_key_worker_pools = api_key_statuses
+
+            for status in api_key_statuses:
+
+                result.total_errors += status.total_errors
+                result.total_bytes_uploaded += status.total_bytes_uploaded
+
+            # get responses information from worker pools
+            responses = {s.all_responses_successful for s in api_key_statuses}
+            if False in responses:
+                # if there is at least one failed response - a whole result is failed too.
+                result.last_responses_status_info = "Last requests on some workers is not successful, see below for more info."
+            elif None in responses:
+                # if there is at least one unknown response - a whole result is unknown too.
+                # NOTE: As the above condition implies, the result will be unknown unly if there is no failed responses.
+                result.last_responses_status_info = None
+            else:
+                result.last_responses_status_info = "All successful"
+
+            # Get health checks the same way as with responses.
+            health_checks = {s.all_health_checks_good for s in api_key_statuses}
+            if False in health_checks:
```

Comment:
Why isn't `result.health_check_result` set here?

Reply:
Because we also do a health check of the copying manager thread itself, and we collect all possible error messages in the `all_health_check_error_messages` list, and only then we set an overall message to the `result.health_check_result`